website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,229 @@
+    },
+    preNavigationHooks: [
+        (crawlingContext) => {
+            const { request, session } = crawlingContext;
+            request.headers = {
+                ...request.headers,
+                ...session.userData?.headers,
+
+            };
+        },
+    ],
+    proxyConfiguration,
+```
+
+Finally in the request handler we make the call using the headers and make sure how many calls are needed to fetch all the data handling pagination.
+
+```js
+async requestHandler(context) {
+        const { log, request, json } = context;
+        const { userData } = request;
+        const { itemsCounter = 0, resultsLimit = 0 } = userData;
+        if (!json.data) {
+            throw new Error('BLOCKED');
+        }
+        const { data } = json;
+        const items = data.list;
+        const counter = itemsCounter + items.length;
+        const dataItems = items.slice(0, resultsLimit && counter > resultsLimit ? resultsLimit - itemsCounter : undefined);
+        await Actor.pushData(dataItems);
+        const { pagination: { page, total } } = data;
+        log.info(`Scraped ${dataItems.length} results out of ${total} from search page ${page}`);
+        const isResultsLimitNotReached = counter < Math.min(total, resultsLimit);
+        if (isResultsLimitNotReached && data.pagination.has_more) {
+            const nextUrl = new URL(request.url);
+            nextUrl.searchParams.set('page', page + 1);
+            await crawler.addRequests([{
+                url: nextUrl.toString(),
+                headers: request.headers,
+                userData: {
+                    ...request.userData,
+                    itemsCounter: itemsCounter + dataItems.length,
+                },
+            }]);
+        }
+    },
+});
+```
+
+One important thing to note here is that we are making this code in a way that we can make any numbers of API calls. 
+
+In this particular example we just made one request and a single session, but you can make more if you need. When the first API call will be completed, it will create the second API call. Again, you can make more calls if needed, but we stopped at two.
+
+To make things more clear, here is how code flow looks:
+
+
+![code flow](./img/code-flow.webp)
+
+## Conclusion
+
+This approach helps us to get a third way to extract the authentication data without actually using browser and pass the data to CheerioCrawler. This significantly improves the performance, reduces the RAM requirement by 50%, and while browser based scraping performance is 10 times slower than pure Cheerio, JSDOM do it just 3-4 times slower, which makes it 2-3 times faster than browser based scraping.
+
+We used our open source library Crawlee in this example, feel free to checkout the project and [give us a star](https://github.com/apify/crawlee). The codebase of the project is already [uploaded here](https://github.com/souravjain540/tiktok-trends), for any queries, [join our discord community](https://apify.com/discord).
```

Comment:
This code isn't runnable easily. It doesn't have a `package.json` and the README doesn't have a "getting started".

```suggestion
The codebase of the project is already [uploaded here](https://github.com/souravjain540/tiktok-trends).
```

Reply:
I would also make the codebase much more prominent, since none of the examples in the blog are runnable.